### ISSUE TITLE
UTF-8 support from ByteString

### ISSUE THREAD

**Issue by hth313** [mrkkrp/megaparsec#194]
It seems to me that UTF-8 encoded ByteString does not work. It `uncons` using the ByteString.Char8, meaning it just takes out a byte. It should perhaps use `uncons` from the utf8-string library?

I do not know the best way to approach this, making megaparsec depend on utf8-string, which also means that it assumes that the ByteString contains UTF-8, which would break anyone that assumes it to be some other (non-ASCII) byte encoded type, like ISO-latin-1.

**Comment by mrkkrp** [comment]
It looks like you want to work with `Text` instead.

**Comment by mrkkrp** [comment]
Token type for `ByteString` should really be `Word8`, not `Char`. It works in terms of bytes.

**Comment by hth313** [comment]
OK, thanks for the advice.

For now I really need to stay with ByteStrings. Part of the reason is that the data comes from C, using some very specific hacks to make it appear as a lazy `ByteString`. I also have the outgoing data that uses builder features of ByteString that seems not to be available in Text.

It is  probably possible to change to Text, but seems a little bit too much at this point and I am not convinced it will be worth the trouble.

I can see two ways to work around it, either convert to String before feeding it to the parser (this was what I did in the past), or use a `newtype` wrapper on the `ByteString`.

**Comment by mrkkrp** [comment]
If you can convert it to `String` then why not to apply `decodeUtf8` to it?

**Comment by hth313** [comment]
I think that would be an extra conversion step. The data comes in a reasonably large UTF-8 `ByteString`, the majority is removed in a preprocessing step before feeding most of the remaining into the parser. From the parser about 20%-25% of it is going to be `String` literals used for various purposes.

I tried to postpone the conversion to `String` by doing it while the parser consumed characters. Given how this and related data is used internally, it may be actually be better to do the conversion one step earlier, as what I did introduced some rough edges in my code base.
